### PR TITLE
Add a generic SSO login API

### PR DESCRIPTION
**Pull request by turt2live**
Rendered: see 'docs' status check
Proposal: https://github.com/matrix-org/matrix-doc/pull/1721

----

This creates a new SSO module which has a generic version of the CAS spec, as per the MSC. CAS was additionally made a subset of SSO given how tightly knit it is: future support for other SSO protocols would be expected to do the same thing.

Large parts of the CAS documentation are copy/pasted from the previous module with minor edits made for clarity or to tie them in better with the SSO module. Similarly, the SSO specification is heavily inspired by the CAS documentation.

### CODE REVIEW

**Comment by turt2live** [comment]
I've requested review from anyone on the team, but would appreciate particular review from @richvdh to ensure accuracy to the proposal and from 1 other member for general review.

File: specification/modules/sso_login.rst [hunk 1]
```diff
@@ -12,14 +12,97 @@
...
+
+2. The homeserver responds with an HTTP redirect to the SSO user interface,
+   which the browser follows.
+
+3. The SSO system authenticates the user.
+
+4. The SSO server and the homeserver interact to verify the user's identity
+   and other authentication information, potentially using a number of redirects.
+
+7. The Matrix client receives the login token and passes it to the |/login|_
```
[the remark this answers is not in the document]
Comment (turt2live) [reply specification/modules/sso_login.rst]: In my defense: counting is hard.

File: specification/modules/sso_login.rst [hunk 2]
```diff
@@ -12,14 +12,97 @@
...
+   Maybe we should recommend that the ``redirectUrl`` should contain a CSRF
+   token which the client should then check before sending the login token to
+   ``/login``?
+
+{{sso_login_redirect_cs_http_api}}
+
+Server behaviour
+----------------
+
+The URI for the SSO system to be used should be configured on the server by the
```
[the remark this answers is not in the document]
Comment (turt2live) [reply specification/modules/sso_login.rst]: I ended up removing it entirely, referencing r0.4.0 for how it *could* be implemented.

**Comment by turt2live** [comment]
as per https://github.com/matrix-org/matrix-doc/pull/1789/commits/aeb524ef89f2ef6751dd0b7067c225b0c8a6a7c2, this now removes CAS from the spec to prefer the generic SSO module. CAS (and `m.login.cas`) is referenced as a possible way to implement SSO through linking to the r0.4.0 spec which describes the required endpoints and flow very well.

File: specification/modules/sso_login.rst [hunk 3]
```diff
@@ -12,14 +12,97 @@
...
+
+If the generated user identifier represents a new user, it should be registered
+as a new user.
+
+Finally, the server should generate a short-term login token. The generated
+token should be a macaroon, suitable for use with the ``m.login.token`` type of
+the |/login|_ API, and `token-based interactive login <#token-based>`_. The
+lifetime of this token SHOULD be limited to around five seconds.
+
+
```
[the remark this answers is not in the document]
Comment (richvdh) [reply specification/modules/sso_login.rst]: well done for decoding this incoherent babble, @turt2live .

File: changelogs/client_server/newsfragments/1789.feature [hunk 4]
```diff
@@ -0,0 +1 @@
+Add a generic SSO login API.
```
Comment (richvdh) [inline changelogs/client_server/newsfragments/1789.feature]: as an aside, I wish we could decide whether changelog entries should have a full-stop or not. 50/50 on https://matrix.org/docs/spec/client_server/unstable.html#changelog at the moment...

File: specification/modules/sso_login.rst [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+.. WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+.. See the License for the specific language governing permissions and
+.. limitations under the License.
+
+SSO client login
+================
+
+.. _module:sso_login:
+
+Single Sign-On (SSO) is a generic web-based protocol for logging in users.
```
Comment (richvdh) [inline specification/modules/sso_login.rst]: I think calling it a protocol is a bit strong. How about: "Single Sign-On (SSO) is a generic term which refers to protocols which allow users to log into applications via a single web-based authentication portal. Examples include "Central Authentication Service" (CAS) and SAML."

File: specification/modules/sso_login.rst [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+As a special case, the Matrix specification supports `CAS-based login <#cas-based-client-login>`_
```
Comment (richvdh) [inline specification/modules/sso_login.rst]: this now seems to be a broken link, and should probably go away.

File: specification/modules/sso_login.rst [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+3. The SSO system authenticates the user.
+
+4. The SSO server and the homeserver interact to verify the user's identity
+   and other authentication information, potentially using a number of redirects.
+
+5. The browser is directed to the ``redirectUrl`` provided by the client with
+   a ``loginToken`` query parameter for the client to log in with.
+
+.. Note::
+   In the older `r0.4.0 version <https://matrix.org/docs/spec/client_server/r0.4.0.html#cas-based-client-login>`_
```
Comment (richvdh) [inline specification/modules/sso_login.rst]: perfect

File: changelogs/client_server/newsfragments/1789.feature (hunk 4 above)
Comment (turt2live) [reply changelogs/client_server/newsfragments/1789.feature]: they should contain a full stop according to our style guide on lists. I'll make a separate PR to fix/clarify this.

Source: https://github.com/matrix-org/matrix-doc/blob/master/meta/documentation_style.rst#stylistic-notes

**Pull request merged**